### PR TITLE
hsi_fits2map function to extract single image maps from image datacube

### PR DESCRIPTION
**Pull request by Laubeee**

### CODE REVIEW

**Comment by nabobalis** [comment]
Any chance you could add a unit test for this as well as a changelog file?

**Comment by Laubeee** [comment]
Sure. About tests: I think that'd mean to have an example-data-cube available... as far as I can see there is neither one in sunkit-instruments nor sunpy? (I didn't look at closer `hsi_image_20101016_191218.fits` yet, but from its name I assume its just one image, not a cube.) Shall I just add one?

**Comment by nabobalis** [comment]
> Sure. About tests: I think that'd mean to have an example-data-cube available... as far as I can see there is neither one in sunkit-instruments nor sunpy? (I didn't look at closer `hsi_image_20101016_191218.fits` yet, but from its name I assume its just one image, not a cube.) Shall I just add one?
> If so, should it rather be a "good showcase" or more of a "minimal example"?

For a test file, I would lean towards a minimal example that is as small as possible. 
It should go in https://github.com/sunpy/sunkit-instruments/tree/master/sunkit_instruments/data/test.

**Comment by kakirastern** [comment]
Thank you for your contribution @Laubeee!

However, it maybe better for you to start a new branch instead of using your `master` branch for the PR? Otherwise you will have no "homebase" to return to figuratively speaking. For best practice please refer to for example:

1. https://docs.astropy.org/en/latest/#developer-documentation
2. https://docs.sunpy.org/en/latest/dev_guide/index.html

This entails starting a new PR with a brief description of what has been introduced or implemented in this repository.

**Comment by kakirastern** [comment]
And also please open the PR against `sunkit-instruments:master` instead of `sunpy:master`, otherwise it will not be able to be merged by the repo maintainer(s) here.

**Comment by kakirastern** [comment]
> And also please open the PR against `sunkit-instruments:master` instead of `sunpy:master`, otherwise it will not be able to be merged by the repo maintainer(s) here.

But I checked it looks alright for me in this regard...

**Comment by nabobalis** [comment]
> Thank you for your contribution @Laubeee!
> 
> However, it maybe better for you to start a new branch instead of using your `master` branch for the PR? Otherwise you will have no "homebase" to return to figuratively speaking. For best practice please refer to for example:
> 
> 1. [docs.astropy.org/en/latest/#developer-documentation](https://docs.astropy.org/en/latest/#developer-documentation)
> 2. [docs.sunpy.org/en/latest/dev_guide/index.html](https://docs.sunpy.org/en/latest/dev_guide/index.html)
> 
> This entails starting a new PR with a brief description of what has been introduced or implemented in this repository.

While I do agree, working off your own fork's master is perfectly fine. 

> And also please open the PR against `sunkit-instruments:master` instead of `sunpy:master`, otherwise it will not be able to be merged by the repo maintainer(s) here.

I think this is a GitHub issue, this PR is opened against `sunkit-instruments:master`.

**Comment by kakirastern** [comment]
Yeah, GitHub keep on accumulating bugs recently for some reason... 

@Laubeee In case in the future you do consider creating a new branch for your future PRs, do consider doing the following locally:
```
git checkout -b <new-branch-name>
```
from the root of the branch you want to start against.

**Comment by kakirastern** [comment]
I am making this suggestion on the strength that you may need to use `rebase` for some repositories in the future, like Astropy. They have a more strict policy due to logging issues.

**Comment by Laubeee** [comment]
Thanks for the suggestions, I'll keep it in mind for next time.
Can this be assigned for review?
btw, any ideas why tox fails? output log is not really helpful.

**Comment by nabobalis** [comment]
> Can this be assigned for review?

I will see who I can get to review this.

> btw, any ideas why tox fails? output log is not really helpful.

The [python_codestyle](https://dev.azure.com/sunpy/sunkit-instruments/_build/results?buildId=11767&view=logs&jobId=4cdc857c-0d48-50f8-2772-3b048c8bb2f3&j=4cdc857c-0d48-50f8-2772-3b048c8bb2f3&t=fe62049c-f9e9-5b31-8bb6-a78aae0c6ebe) check failed for a few reasons:

```
flake8...................................................................Failed
- hook id: flake8
- duration: 1.28s
- exit code: 1

sunkit_instruments/rhessi/tests/test_rhessi.py:200:1: W293 blank line contains whitespace
sunkit_instruments/rhessi/rhessi.py:373:1: W293 blank line contains whitespace
sunkit_instruments/rhessi/rhessi.py:376:1: W293 blank line contains whitespace
```
and
```
Fix End of Files.........................................................Failed
- hook id: end-of-file-fixer
- duration: 0.05s
- exit code: 1
- files were modified by this hook

Fixing changelog/35.feature.rst
```

You can fix these manually or use the pre-commit tools to do it for you.

File: changelog/35.feature.rst
Comment (nabobalis) [inline changelog/35.feature.rst:1]:
```suggestion
Add `sunkit_instruments.rhessi.hsi_fits2map` function to extract `sunpy.map.Map` objects from a RHESSI 4D image cube.
```

File: sunkit_instruments/rhessi/rhessi.py
Comment (nabobalis) [inline sunkit_instruments/rhessi/rhessi.py:363]:
```suggestion
    image_datacube : `str`
```

File: sunkit_instruments/rhessi/rhessi.py
Comment (nabobalis) [inline sunkit_instruments/rhessi/rhessi.py:359]: Is it worth adding a check to ensure that the input is 4D?

Comment (Laubeee) [reply sunkit_instruments/rhessi/rhessi.py:359]: I have not (yet) encountered 2 or 3D hessi-datacubes tbh, but the [original function](http://www.heliodocs.com/xdoc/xdoc_print.php?file=$SSW/hessi/idl/gen/hsi_fits2map.pro) does handle them... a check and a respective message sure wouldn't hurt. If anyone has a use case and sample files I could even implement it

Comment (kakirastern) [reply sunkit_instruments/rhessi/rhessi.py:359]: Try studying this codebase and check maybe use the `WCS` object converted from the FITS file to ascertain... 

Some maybe too-much-for-now references (mainly Papers I and II): https://fits.gsfc.nasa.gov/fits_wcs.html

Comment (kakirastern) [reply sunkit_instruments/rhessi/rhessi.py:359]: We normally do the conversion with some `astropy` magic: https://docs.astropy.org/en/stable/wcs/

File: sunkit_instruments/rhessi/rhessi.py
Comment (hayesla) [inline sunkit_instruments/rhessi/rhessi.py:374]: Just following on from @nabobalis  comment above about checking for 4d cube - you could check the `NAXIS` keyword in the f[0]. I would also add in a check to make sure its a RHESSI imagecube, e.g. checking that the `INSTRUME` keyword is `RHESSI`

File: sunkit_instruments/rhessi/rhessi.py
Comment (hayesla) [inline sunkit_instruments/rhessi/rhessi.py:384]: I think there is issues here for when there is an image cube of a time sequence at one energy range. If there is only one energy range, say for example 12-25keV, the `len(f[1].data[0]["energy_axis"]` here is going to to be  `array([12., 25.], dtype=float32)` of length 2 so will have issues indexing below as the `f[0].data.shape is (x, 1, x, x)` So there should be a check here for those cases

File: sunkit_instruments/rhessi/rhessi.py
Comment (kakirastern) [reply sunkit_instruments/rhessi/rhessi.py:359]: Think the test concerned is intended for a usually 4D RHESSI flare image datacube for some reason...

@DanRyanIrish If you have the time could you please elaborate on this for some clarification? Are all RHESSI flare image datacubes supposedly 4D by construction or design? I couldn't find any relevant info on the Internet from the official docs at first glance for some reason.

**Review by hayesla — COMMENTED** [review]
This is great @laubeee! 

I think the main issue at the moment is that it fails for imagecubes that consist of a time sequence of images at one energy, it may also be worth checking if it works for imagecubes consisting of one time and several energies 

In both these cases the data is in an image cube with that axis (either 0 or 1 in the data.shape) is just 1.

File: sunkit_instruments/rhessi/rhessi.py
Comment (hayesla) [reply sunkit_instruments/rhessi/rhessi.py:359]: Yes they will be by design. The RHESSI image cubes consist of cubes in the shape (n_times, n_energy, image_arr, image_arr) created in the RHESSI GUI.

Comment (hayesla) [reply sunkit_instruments/rhessi/rhessi.py:359]: I think incl. the checks below should be good, and the WCS stuff will be handled in the sunpy.map.Map

File: sunkit_instruments/rhessi/rhessi.py
Comment (kakirastern) [reply sunkit_instruments/rhessi/rhessi.py:374]: Oh I see... But then the 4D assumption seems a bit restrictive to be honest.

File: sunkit_instruments/rhessi/rhessi.py
Comment (kakirastern) [reply sunkit_instruments/rhessi/rhessi.py:359]: Cool... And I still use raw `astropy` wcs in lieu of `sunpy.map.Map` for my work. Danny never complained about this though 🤣

File: sunkit_instruments/rhessi/rhessi.py
Comment (hayesla) [reply sunkit_instruments/rhessi/rhessi.py:374]: But this is just a function for RHESSI imagecubes, which will always be in this format, so I think it should be ok. 

A single RHESSI fits file can be passed directly to sunpy.map.Map

Comment (kakirastern) [reply sunkit_instruments/rhessi/rhessi.py:374]: Cool, good to know 👍

File: sunkit_instruments/rhessi/rhessi.py
Comment (Laubeee) [reply sunkit_instruments/rhessi/rhessi.py:368]: I thought a bit about this. The downside is that you always have to check, is it a list or just a single map? Especially when automating / batch processing.
On the other hand, your suggestion seems nicer when looking at data in interactive mode. Now that raises a bit the question if one of those scenarios should be given priority over the other... personally I'd keep it to always be a list. In interactive mode you quickly realize that you have to unpack another layer, while in a script you might not have tested that edge case and run into errors in the middle of processing files.

BTW: when adapting the logic to handle both 1xN and Nx1 cases you can automatically feed a 1x1 cube as well (which I found [does exist](https://hesperia.gsfc.nasa.gov/rhessi_extras/flare_images/2015/09/30/20150930_1307_1309/hsi_20150930_1307_1309.html)) you'd still receive a dict with one entry that contains a list with also just one entry... I think that is sensible behaviour. In interactive mode one would obviously directly feed it into `sunpy.map.Map`, but in a script it's nice if you don't have to create another special case

**Comment by Laubeee** [comment]
CI somehow never ceases to fail my commits ^^

`ERROR collecting .tox/py38-online/lib/python3.8/site-packages/sunkit_instruments/fermi/__init__.py `
caused by
`E   astropy.utils.exceptions.AstropyDeprecationWarning: The private astropy._erfa module has been made into its own package, pyerfa, which is a dependency of astropy and can be imported directly using "import erfa"` which is actually a warning but handeled as an error... can it be ignored?

**Comment by kakirastern** [comment]
> CI somehow never ceases to fail my commits ^^
> 
> `ERROR collecting .tox/py38-online/lib/python3.8/site-packages/sunkit_instruments/fermi/__init__.py `
> caused by
> `E astropy.utils.exceptions.AstropyDeprecationWarning: The private astropy._erfa module has been made into its own package, pyerfa, which is a dependency of astropy and can be imported directly using "import erfa"` which is actually a warning but handeled as an error... can it be ignored?

Weird, from the look of things I think that warning has been treated as an error by CI/CD... Let see if we can figure this one out for you. Might take a while though.

**Comment by nabobalis** [comment]
We treat all `DeprecationWarning` as errors. 

What has happened here is that astropy has updated to 4.2 and changed how you import "_erfa" thus triggering this.

This will need to be patched separately (I will do it now) and then you can rebase the PR after to get the CI passing again.

**Comment by kakirastern** [comment]
> We treat all `DeprecationWarning` as errors.
> 
> What has happened here is that astropy has updated to 4.2 and changed how you import "_erfa" thus triggering this.
> 
> This will need to be patched separately (I will do it now) and then you can rebase the PR after to get the CI passing again.

Awesome, or if you don't get to it by this weekend's end I will make a patch for it. Just need to use `import erfa` in place of something similar for the formerly `astropy` module.

**Comment by nabobalis** [comment]
I am wrong, the problem is actually inside sunpy and not here.

**Comment by kakirastern** [comment]
> I am wrong, the problem is actually inside sunpy and not in sunkit-instr.

No worries, I implicitly and subconsciously assumed that's the case too...

**Comment by kakirastern** [comment]
Patch is currently being handled here: https://github.com/sunpy/sunpy/pull/4690

**Comment by nabobalis** [comment]
I've restarted the CI. It should be fixed.

**Comment by nabobalis** [comment]
The two failing builds (deveps and online need to be fixed separately). They aren't due to this PR.

**Comment by Laubeee** [comment]
Well, is there a fix in progress / is there anyone working on that?

**Comment by nabobalis** [comment]
> Well, is there a fix in progress / is there anyone working on that?

Master was updated a few days ago. If you rebase/merge in master it should be resolved.

**Comment by Laubeee** [comment]
🥳 
thx!

File: sunkit_instruments/rhessi/rhessi.py
Comment (nabobalis) [inline sunkit_instruments/rhessi/rhessi.py:379]: If possible testing this with a non-rhessi file would be ideal.

**Review by nabobalis — APPROVED** [review]
Code is good for me. I will wait for a second review from someone who understands RHESSI.

File: sunkit_instruments/rhessi/tests/test_rhessi.py
Comment (nabobalis) [inline sunkit_instruments/rhessi/tests/test_rhessi.py:226]: Any chance you can match the message just to make sure?

**Review by kakirastern — APPROVED** [review]
Looks really good!

File: sunkit_instruments/rhessi/rhessi.py
Comment (DanRyanIrish) [reply sunkit_instruments/rhessi/rhessi.py:368]: In the future, when sunpy inherits from ndcube, this question would be resolved by using `NDCubeSequence`and/or `NDCollection` objects.  But it has no bearing on this PR.

File: sunkit_instruments/rhessi/rhessi.py
Comment (hayesla) [inline sunkit_instruments/rhessi/rhessi.py:356]:
```suggestion
def hsi_fits2map(rhessi_imagecube_file):
```

File: sunkit_instruments/rhessi/rhessi.py
Comment (hayesla) [inline sunkit_instruments/rhessi/rhessi.py:374]:
```suggestion
    f = sunpy.io.read_file(rhessi_imagecube_file)
```

File: sunkit_instruments/rhessi/rhessi.py
Comment (DanRyanIrish) [reply sunkit_instruments/rhessi/rhessi.py:359]: > Cool... And I still use raw `astropy` wcs in lieu of `sunpy.map.Map` for my work. Danny never complained about this though 🤣

In ndcube, which is a lower level set of functionalities than sunpy, it is better to be more hands on with the astropy WCS infrastructure.  But in sunpy or sunkit-instruments, which are higher level and can depend on sunpy, it's perfectly fine to use the sunpy classes (e.g. `sunpy.map.Map`) that handle the WCS parsing for you.

File: sunkit_instruments/rhessi/rhessi.py
Comment (hayesla) [reply sunkit_instruments/rhessi/rhessi.py:368]: yep - NDCube will make reading these files easier (however there are some other issues with these RHESSI cubes but we'll cross that bridge when we get there :P )

I would still be of opinion of returning a dictionary of `mapsequence` and just `map` if only a single map exists for each energy for example - I've added a suggestion below.  You can still index a mapsequence as a list, but yes I guess there is a case when you pass a single image rhessi map file then you return a single map, but in this case the user should probably just be using `sunpy.map.Map`

File: sunkit_instruments/rhessi/rhessi.py
Comment (hayesla) [inline sunkit_instruments/rhessi/rhessi.py:405]:
```suggestion
        map_list=[]
```

File: sunkit_instruments/rhessi/rhessi.py
Comment (hayesla) [inline sunkit_instruments/rhessi/rhessi.py:409]:
```suggestion
            map_list.append(Map(data[t][e], header))  # extract image Map
        if len(map_list)==1:
            maps[key] = Map(map_list)
        else:
            maps[key] = Map(map_list, sequence=True)
```

**Review by hayesla — APPROVED** [review]
This is great @Laubeee 

I'm still not 100% about returning a dict of lists, and would like to see this return a dict of `sunpy.map.MapSequence` or `sunpy.map.Map`, but happy to debate this more :P

File: sunkit_instruments/rhessi/rhessi.py
Comment (Laubeee) [reply sunkit_instruments/rhessi/rhessi.py:409]: Thanks for the suggestion!
Like @DanRyanIrish I prefer it was always was a sequence...

in practice, looping changes from
```python
for energy_band in maps.keys():
    for m in maps[energy_band]:
        #do stuff
```
to
```python
from sunpy.map import GenericMap
for energy_band in maps.keys():
    for m in [maps[energy_band]] if isinstance(maps[energy_band], GenericMap) else maps[energy_band]:
        #do stuff
```

which is a bit unnecessarily complicated for my taste, but manageable. I can totally live with it too, but, I don't really see the advantage tbh

**Review by ayshih — COMMENTED** [review]
The name of the function should be changed, e.g., to something like `imagecube2map`.  We don't need to carry the `hsi_` prefix since we have namespaces, and `fits` is ambiguous since there are plenty of non-image-cube RHESSI FITS file.  Just add to the docstring a statement that this function is analogous to SSW's `hsi_fits2map.pro` to enable discovery.  (We don't need to propagate the baggage of SSW function names.)

My inclination is that the function should return a `dict` of `MapSequence`, at least by default, so that there is a consistent return regardless of input file.  You can add an optional keyword argument for the function that will ask it to return a `dict` of `Map` instead if that's possible for the given image cube.

**Review by nabobalis — CHANGES_REQUESTED** [review]
Blocking until the comments have been addressed.

**Comment by Laubeee** [comment]
Should all be there now, except the parameter to return only a Map... any (naming?) suggestions on that one? maybe something like `force_sequence` with default `true`?

In the meantime CI fails at `sunkit_instruments.lyra.lyra.remove_lytaf_events_from_timeseries`. This has failed before and after rebasing it was fine (or maybe it was just the rerun that did it..)

**Comment by nabobalis** [comment]
Most likely a transient failure.

**Comment by Laubeee** [comment]
Can you rerun them?

**Comment by nabobalis** [comment]
> Can you rerun them?

I have just now but since you have not added remote tests, it doesn't matter if they fail.

**Comment by Laubeee** [comment]
Ok, so how do we proceed?

@hayesla are you ok with the way it is now? Do you wish to have a flag to control output as a single map, as Albert suggested?

**Comment by nabobalis** [comment]
> Ok, so how do we proceed?
> 
> @hayesla are you ok with the way it is now? Do you wish to have a flag to control output as a single map, as Albert suggested?

To wait for the three people who have reviewed this PR to give it the all clear again and we can merge.

File: sunkit_instruments/rhessi/rhessi.py
Comment (hayesla) [inline sunkit_instruments/rhessi/rhessi.py:359]:
```suggestion
    Currently assumes input to be 4D.
    
    This function is analogous to the `hsi_fits2map.pro` functionality available in SSW. 
```

**Review by hayesla — APPROVED** [review]

**Comment by hayesla** [comment]
Looks good to be @Laubeee - thank you for this work on this!

File: changelog/35.feature.rst
Comment (ayshih) [inline changelog/35.feature.rst:1]:
```suggestion
Add :func:`sunkit_instruments.rhessi.imagecube2map` function to extract `sunpy.map.MapSequence` objects from a RHESSI 4D image cube.
```

**Review by ayshih — APPROVED** [review]
Some fixes to the changelog entry, but otherwise good to go

A keyword option to return `Map` instead of `MapSequence` can be deferred to a future PR.  My name suggestion `simplify_to_3d`.

**Review by nabobalis — APPROVED** [review]

**Comment by nabobalis** [comment]
Thanks for the PR @Laubeee.

**Pull request merged**
